src/pinvoke_api.cpp

```diff
@@ -0,0 +1,861 @@
+    switch (p) {
+    case 0:
+        return ONE_CMPLX;
+    case 1:
+        return I_CMPLX;
+    case 2:
+        return -ONE_CMPLX;
+    case 3:
+        return -I_CMPLX;
+    }
+    return 0;
+}
+
+void apply_controlled_exp(std::vector<complex>& wfn, std::vector<unsigned> const& b, double phi,
+    std::vector<unsigned> const& cs, std::vector<unsigned> const& qs)
+{
+    std::size_t cmask = make_mask(cs);
+
+    if (isDiagonal(b)) {
+        std::size_t mask = make_mask(qs);
+        complex phase = std::exp(complex(0., -phi));
+
+        for (std::intptr_t x = 0; x < static_cast<std::intptr_t>(wfn.size()); x++) {
+            if ((x & cmask) == cmask) {
+                wfn[x] *= (poppar(x & mask) ? phase : std::conj(phase));
+            }
+        }
+    } else { // see Exp-implementation-details.txt for the explanation of the algorithm below
+        std::size_t xy_bits = 0;
+        std::size_t yz_bits = 0;
+        int y_count = 0;
+        for (unsigned i = 0; i < b.size(); ++i) {
+            switch (b[i]) {
+            case PauliX:
+                xy_bits |= (1ull << qs[i]);
+                break;
+            case PauliY:
+                xy_bits |= (1ull << qs[i]);
+                yz_bits |= (1ull << qs[i]);
+                ++y_count;
+                break;
+            case PauliZ:
+                yz_bits |= (1ull << qs[i]);
+                break;
+            case PauliI:
+                break;
+            }
+        }
+
+        real1 alpha = (real1)std::cos(phi);
+        complex beta = (real1)std::sin(phi) * iExp(3 * y_count + 1);
+        complex gamma = (real1)std::sin(phi) * iExp(y_count + 1);
+
+        for (std::intptr_t x = 0; x < static_cast<std::intptr_t>(wfn.size()); x++) {
+            std::intptr_t t = x ^ xy_bits;
+            if (x < t && ((x & cmask) == cmask)) {
+                auto parity = poppar(x & yz_bits);
+                auto a = wfn[x];
+                auto b = wfn[t];
+                wfn[x] = alpha * a + (parity ? -beta : beta) * b;
+                wfn[t] = alpha * b + (parity ? -gamma : gamma) * a;
+            }
+        }
+    }
+}
+
+extern "C" {
+
+/**
+ * (External API) Initialize a simulator ID with 0 qubits
+ */
+MICROSOFT_QUANTUM_DECL unsigned init()
+{
+    META_LOCK_GUARD()
+
+    unsigned sid = simulators.size();
+
+    for (unsigned i = 0; i < simulators.size(); i++) {
+        if (simulators[i] == NULL) {
+            sid = i;
+            break;
+        }
+    }
+
+    QInterfacePtr simulator =
+        CreateQuantumInterface(QINTERFACE_QUNIT, QINTERFACE_QFUSION, QINTERFACE_OPTIMAL, 4, 0, rng);
+    if (sid == simulators.size()) {
+        simulators.push_back(simulator);
+    } else {
+        simulators[sid] = simulator;
+    }
+
+    return sid;
+}
+
+/**
+ * (External API) Destroy a simulator (ID will not be reused)
+ */
+MICROSOFT_QUANTUM_DECL void destroy(_In_ unsigned sid)
+{
+    META_LOCK_GUARD()
+    // SIMULATOR_LOCK_GUARD(sid)
+
+    simulators[sid] = NULL;
```

Comment:
FWIW, memory indicates `simulators[sid].reset();` is slightly more canonical, but I could be wrong about that.

Reply:
That would release ownership, (as does `NULL` assignment,) but we do need to check for empty pointers, for reuse. You might very well be right, but I specifically need a detectable `NULL` upon release.

Reply:
It's a smart pointer. It does that already.